instrumentctl/g9_driver.py

```diff
@@ -0,0 +1,255 @@
+                self.SOTDF = data[17:21]
+                self.SOTSF = data[27:31]
+
+                # Unit status
+                if self.US != b'\x00\x01':
+                    if self.unit_state_error(self.US):
+                        raise ValueError("Error was detected in Unit State but was not identified. Could be more than one")
+
+                # Input Terminal Data Flags (1 - ON 0 - OFF)
+                if not self.check_flags13(self.SITDF):
+                    err = []
+                    gates = self.bytes_to_binary(self.SITDF[-3:])
+                    for i in range(20):
+                        if gates[-i + 1] == "0":
+                            err.append(i)
+                    self.input_flags = gates
+                    raise ValueError(f"An input is either off or throwing an error: {err}")
+
+                # Input Terminal Status Flags (1 - OK 0 - OFF/ERR)
+                if not self.check_flags13(self.SITSF):
+                    # if error dected checkout terminal error cause
+                    if self.safety_in_terminal_error(data[31:55][-10:]):
+                        raise ValueError("Error was detected in inputs but was not found")
+                       
+                # Output Terminal Data Flags (1 - ON 0 - OFF)
+                if not self.check_flags13(self.SOTDF):
+                    err = []
+                    gates = self.bytes_to_binary(self.SOTDF[-3:])
+                    for i in range(NUMIN):
+                        if gates[-i + 1] == "0":
+                            err.append(i)
+                    raise ValueError(f"There is output(s) off: {err}")
+                
+                # Output Terminal Status Flags (1 - OK 0 - OFF/ERR)
+                if not self.check_flags13(self.SOTSF):
+                    if self.safety_out_terminal_error(data[55:71][-10:]):
+                        raise ValueError("Error was detected in outputs but was not found")
+                    
+                # Optional Communication data 
+                OCTD = data[7:11]
+                if OCTD != SNDDATA:
+                    raise ValueError("Optional Transmission data doesn't match data sent to the G9SP")
+                
+                # # TODO: Need to add error log
+                # errorLog = data[108:149]
+
+                # # TODO: Need to add operation log
+                # operationLog = data[148:199]
+                    
+            else:
+                raise ValueError("Response length was not OxC3, either an error or command formate invalid.")
+
+    """
+    0: No error
+    1: Invalid configuration
+    2: External test signal failure
+    3: Internal circuit error
+    4: Discrepancy error
+    5: Failure of the associated dual-channel input
+    """
+    # checks all the SITSFs, throws error is one is found
+    def safety_in_terminal_error(self, data):
+        if len(data) != 10:
+            raise ValueError(f"Expected 24 bytes, but received {len(data)}.")
```

Comment:
confusing message here. `len(data) != 10` but then says `Expected 24 bytes`. Shouldn't this match the actual expected lengths?

Reply:
That is a typo. The 'Safety Input Terminal Error Causes' section of the data is 24 bytes but only 10 bytes indicate the error causes whilst the remaining are reserved. We accidentally put in 24 instead of 10.